daemon/main.cpp

```diff
@@ -1884,10 +1884,15 @@ int Daemon::answer_client_requests()
             }
 
             for (map<string, NativeEnvironment>::iterator it = native_environments.begin();
-                 it != native_environments.end(); ++it) {
+                 it != native_environments.end(); ) {
                 if (it->second.create_env_pipe && FD_ISSET(it->second.create_env_pipe, &listen_set)) {
-                    create_env_finished(it->first);
+                    if(!create_env_finished(it->first))
+                    {
+                        native_environments.erase(it++);
+                        continue;
+                    }
                 }
+                ++it;
```

Comment:
Because `erase()` invalidates the iterator:

```c++
   native_environments.erase(it++);
```
`erase()` is called with the value of the current `it`, but is incremented before the next loop iteration since it uses post-increment. If it's not getting erased, then it needs to be incremented anyway, hence moving it out the for loop. You can't call `erase(it)` within the body of the for loop, because then the `it++` operation that happens after each loop iteration will be using an invalidated `it`.